# Week 12: Linear Models

> Laiba Siddique

## Imports

In [1]:
import numpy as np
from scipy import stats
from scipy import special
from pandas import DataFrame

In [9]:
import plotly.graph_objects as go
import plotly.express as px
import plotly.figure_factory as ff
import plotly.io as pio
pio.renderers.default = "vscode"
pio.templates.default = 'plotly_white'

In [3]:
from patsy import dmatrices
import statsmodels.api as sm

## Questions

View the data below, that shows data for $100$ observations. There are two continuous numerical variables and a dependent variable.

In [4]:
# Seed the pseudo-random number generator
np.random.seed(12)

# Sample size
n=100

# Create data
wcc = np.round(np.random.normal(15, 5, n), 0)
crp = (wcc * 2) + np.round(np.random.normal(0, 10, n), 0)

lipase = wcc + crp + np.round(np.random.normal(2, 10, n), 0)

# Pandas dataframe obejct
df = DataFrame(
    {'WCC':wcc,
     'CRP':crp,
     'Lipase':lipase}
)

In [10]:
px.scatter_matrix(
    df,
    title='Scatter plot matrix'
)

ValueError: Mime type rendering requires nbformat>=4.2.0 but it is not installed

Figure({
    'data': [{'dimensions': [{'axis': {'matches': True},
                              'label': 'WCC',
                              'values': {'bdata': ('AAAAAAAAMUAAAAAAAAAoQAAAAAAAAD' ... 'AAAAAgQAAAAAAAAChAAAAAAAAAMEA='),
                                         'dtype': 'f8'}},
                             {'axis': {'matches': True},
                              'label': 'CRP',
                              'values': {'bdata': ('AAAAAAAANEAAAAAAAAAoQAAAAAAAAE' ... 'AAAAA0QAAAAAAAACxAAAAAAAAAMEA='),
                                         'dtype': 'f8'}},
                             {'axis': {'matches': True},
                              'label': 'Lipase',
                              'values': {'bdata': ('AAAAAAAARkAAAAAAAAA0QAAAAAAAAE' ... 'AAAIBAQAAAAAAAgERAAAAAAAAAIEA='),
                                         'dtype': 'f8'}}],
              'hovertemplate': '%{xaxis.title.text}=%{x}<br>%{yaxis.title.text}=%{y}<extra></extra>',
              'legendgroup': '',
              'marker': {'color': '#636efa', 'symbol': 'circle'},
              'name': '',
              'showlegend': False,
              'type': 'splom'}],
    'layout': {'dragmode': 'select',
               'legend': {'tracegroupgap': 0},
               'template': '...',
               'title': {'text': 'Scatter plot matrix'}}
})

In [11]:
# Target vector and design matrix
y, X = dmatrices('Lipase ~ WCC + CRP', data = df)

### Question 1: Create multiple linear regression model with `WCC` and `CRP` as predictors of `Lipase`.

In [12]:
# 1
model = sm.OLS(y, X).fit()

model.summary2()

<class 'statsmodels.iolib.summary2.Summary'>
"""
                 Results: Ordinary least squares
=================================================================
Model:              OLS              Adj. R-squared:     0.759   
Dependent Variable: Lipase           AIC:                754.2907
Date:               2026-03-31 17:42 BIC:                762.1062
No. Observations:   100              Log-Likelihood:     -374.15 
Df Model:           2                F-statistic:        157.3   
Df Residuals:       97               Prob (F-statistic): 3.61e-31
R-squared:          0.764            Scale:              107.28  
-------------------------------------------------------------------
            Coef.    Std.Err.      t      P>|t|     [0.025   0.975]
-------------------------------------------------------------------
Intercept   2.3741     3.0129    0.7880   0.4326   -3.6056   8.3538
WCC         0.5426     0.2684    2.0213   0.0460    0.0098   1.0754
CRP         1.1827     0.1014   11.6636   0.0000    0.9814   1.3839
-----------------------------------------------------------------
Omnibus:              0.113        Durbin-Watson:           1.936
Prob(Omnibus):        0.945        Jarque-Bera (JB):        0.287
Skew:                 0.029        Prob(JB):                0.866
Kurtosis:             2.744        Condition No.:           96   
=================================================================
Notes:
[1] Standard Errors assume that the covariance matrix of the
errors is correctly specified.
"""

### Question 2: Comment on the *F* statistic.

The *F* statistic is large at $157.3$, with a small *p* value of $0$. There is enough evidence at the $5\%$ level of significance to show that at least one of the indpendent variables is linearly associated with the dependent variable.

### Question 3: Comment on the individual coefficients and their *p* values.

Only the coefficient for `WCC` has a small *p* value. There is enough evidence at the $5\%$ level of significance to show that `WCC` is linearly associated with `Lipase` after correcting for `CRP`.

### Question 4: Comment on the $R^{2}$ value.

The model explains $76.4$ as the variability in the `Lipase` levels